networkx/algorithms/isomorphism/vf2pp_helpers/state.py

```diff
@@ -0,0 +1,225 @@
+from networkx.algorithms.isomorphism.vf2pp_helpers.candidates import _find_candidates
+
+
+def _update_Tinout(new_node1, new_node2, graph_params, state_params):
+    """Updates the Ti/Ti_out (i=1,2) when a new node pair u-v is added to the mapping.
+
+    Notes
+    -----
+    This function should be called right after the feasibility checks are passed, and node1 is mapped to node2. The
+    purpose of this function is to avoid brute force computing of Ti/Ti_out by iterating over all nodes of the graph
+    and checking which nodes satisfy the necessary conditions. Instead, in every step of the algorithm we focus
+    exclusively on the two nodes that are being added to the mapping, incrementally updating Ti/Ti_out.
+
+    Parameters
+    ----------
+    new_node1, new_node2: int
+        The two new nodes, added to the mapping.
+
+    graph_params: namedtuple
+        Contains all the Graph-related parameters:
+
+        G1,G2: NetworkX Graph or MultiGraph instances.
+            The two graphs to check for isomorphism or monomorphism
+
+        G1_labels,G2_labels: dict
+            The label of every node in G1 and G2 respectively
+
+    state_params: namedtuple
+        Contains all the State-related parameters:
+
+        mapping: dict
+            The mapping as extended so far. Maps nodes of G1 to nodes of G2
+
+        reverse_mapping: dict
+            The reverse mapping as extended so far. Maps nodes from G2 to nodes of G1. It's basically "mapping" reversed
+
+        T1, T2: set
+            Ti contains uncovered neighbors of covered nodes from Gi, i.e. nodes that are not in the mapping, but are
+            neighbors of nodes that are.
+
+        T1_out, T2_out: set
+            Ti_out contains all the nodes from Gi, that are neither in the mapping nor in Ti
+    """
+    G1, G2, _, _, _, _, _ = graph_params
+    mapping, reverse_mapping, T1, T1_out, T2, T2_out = state_params
+
+    uncovered_neighbors_G1 = {nbr for nbr in G1[new_node1] if nbr not in mapping}
+    uncovered_neighbors_G2 = {
+        nbr for nbr in G2[new_node2] if nbr not in reverse_mapping
+    }
+
+    # Add the uncovered neighbors of node1 and node2 in T1 and T2 respectively
+    T1.discard(new_node1)
+    T2.discard(new_node2)
+    T1.update(uncovered_neighbors_G1)
+    T2.update(uncovered_neighbors_G2)
```

Comment:
I think this might get tripped up by a selfloop from `new_node1` to itself. We discard the `new_node1` and then in the update we add it again if there is a selfloop.

If we update before we discard then we will certainly remove it from the set. 
Can you construct a test which breaks with the current code? Then I think the solution is to `discard` after all the updates are done.